### ISSUE TITLE
`*type` and `**type` is not parsed properly for Python 2 type hints

### ISSUE THREAD

**Issue by skasch** [microsoft/pylance-release#1191]
## Environment data

-   Language Server version: 2021.4.2
-   OS and version: Ubuntu 16.04
-   Python version (& distribution if applicable, e.g. Anaconda): 2.7.12

## Expected behaviour

`*type` and `**type` type hints are parsed properly by Pylance.

## Actual behaviour

`*type` and `**type` type hints raise an error.

## Logs

I did not manage to get any relevant log, even with `Trace`-level logging enabled.

## Code Snippet / Additional information

```python
def test_fun(*args, **kwargs):
    # type: (*int, **float) -> int
    return sum(args) + sum(round(kwarg) for kwarg in kwargs.values())
```

According to [PEP-484](https://www.python.org/dev/peps/pep-0484/#suggested-syntax-for-python-2-7-and-straddling-code), this should be the standard way to type hint `*` and `**` arguments in older versions of Python. However, writing the type hint that way raises errors:

![image](https://user-images.githubusercontent.com/8453072/116053283-ca964a00-a67a-11eb-836a-a602f7e7ceb9.png)

In the code, `args` is considered to have type `Tuple[Any, ...]` and `kwargs` is considered to have type `Dict[str, Any]`.

However, it looks like using line-by-line type hints work:

```python
def test_fun(
    *args,  # type: int
    **kwargs,  # type: float
):
    # type: (...) -> int
    return sum(args) + sum(round(kwarg) for kwarg in kwargs.values())
```

Here, `args` is considered to have type `Tuple[int, ...]`, and `kwargs` is considered to have type `Dict[str, float]`.

I also tried removing the `*` and `**`:

```python
def test_fun(*args, **kwargs):
    # type: (int, float) -> int
    return sum(args) + sum(round(kwarg) for kwarg in kwargs.values())
```

In that case, Pylance does find the right types, `args` is considered to have type `Tuple[int, ...]` and `kwargs` to have type `Dict[str, float]`, but this is inconsistent with the expected format (and I find it also harder to read).

**Comment by erictraut** [comment]
Pylance is designed for Python 3 and doesn't generally support Python 2. It does have minimal support for function annotation comments with the goal of supporting type stubs for libraries that work with Python 2 and 3, but this support has many limitations.

I'm curious why you are still using function annotation comments in your code. Is it not feasible to switch to Python 3 syntax?

We don't plan to make any further significant investments in Python 2 function annotation comments. However, in this particular case, it's simple to eliminate the parsing error that you're seeing (just a one-line change), so I'll make that change in the parser. I'm not adding full semantic validation (e.g. verifying that `*` and `**` parameter annotations line up with the corresponding `*args` and `**kwargs` parameters in the signature). That would require a much bigger investment.

The parser change will be included in the next release of pylance.

**Comment by skasch** [comment]
Thank you very much for taking the time to fix that, despite it being a Python 2 issue specifically!

Unfortunately, I'm currently working on a pretty large codebase running in Python 2, which is why I cannot use the Python 3 syntax. We are working on a transition to Python 3, but in the meantime, we try as much as possible to write Python 2 code that can also be interpreted as Python 3 code. Thanks to that, I've found that Pylance is really useful for me to analyze the code, despite it being Python 2 code. There are of course some limitations, and having to use type comments instead of type annotations is one of them.

**Comment by jakebailey** [comment]
This issue has been fixed in version 2021.4.3, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202143-29-april-2021
